libcfnet/file_stream.c

```diff
@@ -338,6 +338,194 @@ static bool ProtocolSendError(SSL *conn, bool flush, const char *fmt, ...)
+ * @param in_buf The input buffer
+ */
+static void MoveLeftoversToFrontOfInputBuffer(rs_buffers_t *bufs, char *in_buf)
+{
+    assert(bufs != NULL);
+    assert(in_buf != NULL);
+    assert(bufs->next_in >= in_buf);
+
+    if (bufs->avail_in > 0)
+    {
+        memmove(in_buf, bufs->next_in, bufs->avail_in);
+    }
+    bufs->next_in = in_buf;
+}
+
+/**
+ * @brief Fill input buffer with messages received from remote host
+ *
+ * @param bufs RS buffers
+ * @param in_buf Input buffers
+ * @param conn SSL connection
+ * @return false in case of failure
+ */
+static bool FillInputBufferFromHost(
+    rs_buffers_t *bufs, char *in_buf, SSL *conn)
+{
+    assert(bufs != NULL);
+    assert(in_buf != NULL);
+    assert(conn != NULL);
+
+    MoveLeftoversToFrontOfInputBuffer(bufs, in_buf);
+
+    if (bufs->eof_in != 0)
+    {
+        /* No more data to fill the buffer with */
+        return true;
+    }
+
+    if (bufs->avail_in > PROTOCOL_MESSAGE_SIZE)
+    {
+        /* We don't have space for another message */
+        return true;
+    }
+
+    size_t msg_len;
+    bool eof;
+    if (!ProtocolRecvMessage(
+            conn, bufs->next_in + bufs->avail_in, &msg_len, &eof))
+    {
+        return false;
+    }
+
+    bufs->eof_in = eof ? 1 : 0;
+    bufs->avail_in += msg_len;
+    return true;
+}
+
+/**
+ * @brief Fill input buffer with contents from file
+ *
+ * @param bufs RS buffers
+ * @param in_buf Input buffer
+ * @param file The file
+ * @return false in case of failure
+ */
+static bool FillInputBufferFromFile(
+    rs_buffers_t *bufs, char *in_buf, FILE *file)
+{
+    assert(bufs != NULL);
+    assert(in_buf != NULL);
+    assert(file != NULL);
+
+    MoveLeftoversToFrontOfInputBuffer(bufs, in_buf);
+
+    if (bufs->eof_in != 0)
+    {
+        /* No more data to fill the buffer with */
+        return true;
+    }
+
+    assert(bufs->avail_in <= PROTOCOL_MESSAGE_SIZE);
+    const size_t remaining = PROTOCOL_MESSAGE_SIZE - bufs->avail_in;
+    if (remaining == 0)
+    {
+        /* There is no more space in buffer */
+        return true;
+    }
+
+    size_t num_bytes_read =
+        fread(bufs->next_in + bufs->avail_in, 1, remaining, file);
+    if ((num_bytes_read == 0) && ferror(file))
+    {
+        /* Failed to read */
+        return false;
+    }
+
+    bufs->eof_in = feof(file);
+    bufs->avail_in += num_bytes_read;
+    return true;
+}
+
+/**
+ * @brief Send contents of output buffer to remote host
+ *
+ * @param bufs RS buffers
+ * @param out_buf Output buffer
+ * @param is_done Whether to set End-of-File flag
+ * @param conn SSL connection
+ * @return false in case of failure
+ */
+static bool DrainOutputBufferToHost(
+    rs_buffers_t *bufs, char *out_buf, bool is_done, SSL *conn)
+{
+    assert(bufs != NULL);
+    assert(out_buf != NULL);
+    assert(conn != NULL);
+
+    const size_t num_bytes = bufs->next_out - out_buf;
+    assert(num_bytes <= PROTOCOL_MESSAGE_SIZE);
+    if ((num_bytes == 0) && !is_done)
```

Comment:
like the above section where you check == 0 I wonder if num_bytes could ever be negative?

Reply:
`next_out` references a pointer which on entry points to the start of the output
buffer (`out_buf`). Updated to point to the byte after the last one filled. Hence `next_out` is always greater or equal to `out_buf`. Thus, num_bytes can never be less than zero. Assert above catches it if mistakes where made.

Reply:
The Stream API of librsync is quite confusing, given that the attributes of the `rs_buffers_t` struct means different things before and after a job iteration.